pkg/util/cache.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package util
+
+import (
+	"sync"
+	"time"
+)
+
+// CacheItem wraps T so it has a LastUpdated field and can be used with Cache.
+type CacheItem[T any] struct {
+	// Item is the cached item.
+	Item T
+
+	// LastUpdated is the last time the item was updated.
+	LastUpdated time.Time
+}
+
+// Cache is a generic implementation of a cache that can be configured with a
+// maximum number of items and can evict items after a certain amount of time.
+type Cache[T any] struct {
+	mu       sync.Mutex
+	items    map[string]CacheItem[T]
+	maxItems int
+	done     chan struct{}
+	donce    sync.Once
+
+	// expiredChan returns a channel on which the IDs of items that are expired
+	// are sent. This channel is closed when the cache is closed, but only after
+	// all, pending eviction notices are received.
+	expiredChan chan string
+	expiringNum sync.WaitGroup
+}
+
+// NewCache initializes a new cache with the provided expiration options.
+func NewCache[T any](
+	expireAfter, checkExpireInterval time.Duration,
+	maxItems int) *Cache[T] {
+
+	c := &Cache[T]{
+		expiredChan: make(chan string),
+		done:        make(chan struct{}),
+		items:       map[string]CacheItem[T]{},
+		maxItems:    maxItems,
+	}
+
+	// Start a goroutine to expire items from the cache.
+	go func() {
+		ticker := time.NewTicker(checkExpireInterval)
+
+		// evictFn iterates over the cached items and evicts expired items
+		evictFn := func() {
+			c.mu.Lock()
+			defer c.mu.Unlock()
+			for k, v := range c.items {
+				if time.Since(v.LastUpdated) > expireAfter {
+					// Delete the item from the cache.
+					delete(c.items, k)
+
+					// Notify subscribers that the item with the ID of k has
+					// been evicted if the cache is not closed. Otherwise we
+					// add one to the number of items being expired. This
+					// ensures the channel cannot be closed until all, pending
+					// notifications are sent.
+					select {
+					case <-c.done:
+					default:
+						c.expiringNum.Add(1)
+						go func(k string) {
+							c.expiredChan <- k
+							c.expiringNum.Done()
+						}(k)
+					}
+				}
+			}
+		}
+		for {
+			select {
+			case <-c.done:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				evictFn()
+			}
+		}
+	}()
+
+	return c
+}
+
+// ExpiredChan returns a channel on which the IDs of items that are expired are
+// sent. This channel is closed when the cache is closed, but only after all,
+// pending eviction notices are received.
+func (c *Cache[T]) ExpiredChan() <-chan string {
```

Comment:
This is only called in the tests, right? Does that mean that the send done as part of the expire is going to block because there is nothing reading from the channel?

Reply:
I don't think it's blocking anything as the expire send is always done in a goroutine. With that being said, the newly added `InitOvfCacheAndLockPool` will now read from this channel to clean up the corresponding lock item.

Reply:
The goroutine will block because there is/was nothing reading from the channel.